packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -0,0 +1,203 @@
+
+import "../../lib/Claimable.sol";
+import "../../lib/ERC20SafeTransfer.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/ReentrancyGuard.sol";
+import "../../lib/SignatureUtil.sol";
+import "../../lib/EIP712.sol";
+
+
+/// @title Fast withdrawal agent implementation.
+///
+///        A mediator/liquidity provider (whoever provides the funds immediately) can call
+///        `executeFastWithdrawals` to provide users immediately with funds onchain.
+///        This allows the security of the funds to be handled by any EOA or smart contract.
+///
+///        Users that want to make use of this functionality have to
+///        authorize this contract as their agent.
+///
+///        TODO: generalize this (here or another agent) with any function call
+///        a user wants to do to be able to simulate completely onchain smart wallet with
+///        funds stored in the Loopring protocol.
+///
+/// @author Brecht Devos - <brecht@loopring.org>
+contract FastWithdrawalAgent is ReentrancyGuard
+{
+    using AddressUtil       for address;
+    using AddressUtil       for address payable;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using SignatureUtil     for bytes32;
+
+    struct FastWithdrawal
+    {
+        address exchange;
+        address from;
+        address to;
+        address token;
+        uint    amount;
+        address feeToken;
+        uint    fee;
+        uint    validUntil;
+        bool    onchainFeePayment;
+        bytes   signature;
+    }
+
+    // Replay protection
+    mapping (bytes32 => bool) txHash;
+
+    // EIP712
+    bytes32 constant public FASTWITHDRAWAL_TYPEHASH = keccak256(
+        "FastWithdrawal(address exchange,address from,address to,address token,uint256 amount,address feeToken,uint256 fee,bool onchainFeePayment,uint256 validUntil)"
+    );
+    bytes32 public DOMAIN_SEPARATOR;
+
+    constructor()
+        public
+    {
+        //DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("FastWithdrawalAgent", "1.0", address(this)));
+    }
+
+    function executeFastWithdrawals(FastWithdrawal[] memory fastWithdrawals)
+        public
+        payable
+        nonReentrant
+    {
+        // Do all fast withdrawals
+        for (uint i = 0; i < fastWithdrawals.length; i++) {
+            executeFastWithdrawal(fastWithdrawals[i]);
+        }
+        // Return any ETH left into this contract
+        // (can happen when more ETH is sent than needed for the fast withdrawals)
+        msg.sender.sendETHAndVerify(address(this).balance, gasleft());
+    }
+
+    // -- Internal --
+
+    function executeFastWithdrawal(FastWithdrawal memory fastWithdrawal)
+        internal
+    {
+        // Currently here because the compiler hangs when doing this in the constructor
+        if (DOMAIN_SEPARATOR == bytes32(0)) {
+            DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("FastWithdrawalAgent", "1.0", address(this)));
+        }
+
+        // Compute the hash
```

Comment:
maybe we shoud prevent the situation where `fastWithdrawal.from == msg.sender` or `amount == 0`

Reply:
I agree that it's a useless. On the other hand, I don't see much need to test these special cases onchain as it makes no sense to execute these fast withdrawals. I don't see why someone (and these are not normal users, these are executed by people running a business) would pay gas to get these onchain.